erigon-lib/state/state_changeset.go

```diff
@@ -135,26 +101,153 @@ func (d *StateDiffDomain) GetDiffSet() (keysToValue []KVPair) {
 	return d.prevValsSlice
 }
 
-type ChangesetStorage struct {
-	st *lru.Cache[common.Hash, *StateChangeSet]
+func SerializeDiffSet(diffSet []DomainEntryDiff, out []byte) []byte {
+	ret := out
+	// Write a small dictionary for prevStepBytes
+	dict := make(map[string]byte)
+	id := byte(0x00)
+	for _, diff := range diffSet {
+		if _, ok := dict[string(diff.PrevStepBytes)]; ok {
+			continue
+		}
+		dict[string(diff.PrevStepBytes)] = id
```

Comment:
sure that `id` cannot be overflowed by passing `len(diffSet)>2^8`? same with line 117, maybe limit diffSet size in one encoding?

Reply:
ah that is limited by count of domains, i see. but size check would be good anyway

Reply:
I mean, it will overflow if the block processed 255 steps, 1 step is 1.5 mn transactions, so we are good unless gasLimit is infinite